### PR TITLE
feat(fpm): bind regression models to worker roles, change regression features, and modify bucketing dimensions

### PR DESCRIPTION
**Pull request by diffractiongoo**
## Summary

Supersedes [ai-dynamo/aiconfigurator#1602](https://github.com/ai-dynamo/aiconfigurator/pull/1602) after the AIConfigurator repository code freeze. This ports the same reviewed implementation onto AISimulate `main` and records the final AIC synchronization boundary.

This change redesigns the FPM regression fallback around an immutable engine worker role while preserving the existing Native AIC model.

### Worker binding and feature selection

Each regression model is constructed as `prefill`, `decode`, or `aggregated` and owns one two-dimensional regression store. All roles use the common axis order:

$$
x=[\text{critical-rank attention work},\ \text{cross-rank FFN/MoE work}].
$$

For attention-DP rank $d$, define:

- $P_d$ — `sum_prefill_tokens`
- $H_d$ — `sum_prefill_kv_tokens`
- $N_d$ — `num_prefill_requests`
- $B_d$ — `num_decode_requests`
- $K_d$ — `sum_decode_kv_tokens`

$H_d$ and $P_d$ are sums over the Prefill requests scheduled on rank $d$; $K_d$ is a sum over that rank's Decode requests. They are not per-request lengths. $N_d$ and $B_d$ are the corresponding per-rank request counts. Cached Prefill KV contributes only when the rank has fresh Prefill work:

$$
\widetilde H_d=H_d\,\mathbf{1}[P_d>0].
$$

For $P_d>0$, metric validation guarantees $N_d>0$. Under the balanced-request approximation, the total Prefill attention-pair work on rank $d$ is

$$
\begin{aligned}
Q_d
&=N_d\left[
\left(\frac{H_d}{N_d}\right)\left(\frac{P_d}{N_d}\right)
+\frac{(P_d/N_d)(P_d/N_d+1)}{2}
\right]\\
&=\frac{H_dP_d}{N_d}+\frac{P_d^2}{2N_d}+\frac{P_d}{2}.
\end{aligned}
$$

When $P_d=0$, define $Q_d=0$. Thus $Q_d$ already includes the $N_d$ requests on rank $d$ and must not be multiplied by $N_d$ again. $N_d$ is not the attention-DP size.

The three configurable, unity-default weights are:

- $\alpha$ — `regression_attention_kv_weight`, which scales KV-token-related attention work for every role: $\widetilde H$ for Prefill, $K$ for Decode, and $\widetilde H+K$ for Aggregated.
- $\beta$ — `regression_prefill_attention_pair_weight`, which scales only the Prefill estimate $Q$ and does not represent Decode attention.
- $\gamma$ — `regression_ffn_token_weight`, which scales global tokenwise FFN/MoE work: $P$ for Prefill, $B$ for Decode, and $P+B$ for Aggregated.

The selected features are:

$$
x_P=
\left[
\max_d\left(\alpha\widetilde H_d+\beta Q_d\right),
\ \gamma\sum_d P_d
\right],
$$

$$
x_D=
\left[
\alpha\max_d K_d,
\ \gamma\sum_d B_d
\right],
$$

$$
x_A=
\left[
\max_d\left(\alpha(\widetilde H_d+K_d)+\beta Q_d\right),
\ \gamma\sum_d(P_d+B_d)
\right].
$$

These vectors remain two-dimensional for both single-rank and multi-rank engines. For `attention_dp = 1`, the maximum and sum reduce to that rank. For `attention_dp > 1`, critical attention uses a maximum across ranks and global FFN/MoE work uses a sum.

For Aggregated workers, the complete attention score is composed per rank before taking the maximum, so terms from different ranks cannot be combined into an unrealizable critical path. Counters are converted to `f64` before multiplication and cross-rank summation.

Prefill models reject scheduled Decode work, Decode models reject fresh Prefill work, and Aggregated models accept Prefill-only, Decode-only, mixed, and phase-separated-rank iterations. Cached Prefill metadata without fresh tokens remains idle or Decode-compatible.

Native AIC remains role-free and continues to infer Prefill, Decode, or mixed work per iteration using its existing representative-rank features and correction stores.

### Bucketing, retention, and fitting

Each regression model now has one dynamic two-dimensional observation store rather than separate workload-inferred stores. Every accepted observation retains its raw feature vector and observed latency, while bucket placement applies `log1p` to each raw coordinate:

$$
b_j=\log(1+x_j).
$$

With the default `bucket_count=16`, retention uses a $4\times4$ grid. Bounds expand as new extrema arrive and existing observations are rebucketed; bounds do not shrink. When the global observation cap is exceeded, the existing fattest-cell policy evicts an observation from the most populated bucket.

After insertion and eviction, the model refits using only the surviving raw observations—not their logarithmic bucket coordinates. Each raw feature axis is standardized using post-eviction Welford statistics. Near-constant axes are marked inactive, and the remaining axes are fitted with the existing nonnegative active-set linear regression, a free intercept, and slope-only regularization fallback. The resulting fit retains its standardization snapshot for prediction, supports extrapolation, and preserves the positive nonidle prediction floor.

Diagnostics retain their existing schema, but regression readiness and retained-observation counts now describe this single role-bound store.

### Transport and Native tuning parity

The Python facade transports nonfinite values for the three regression weights as exact valid-JSON sentinels—`"NaN"`, `"Infinity"`, and `"-Infinity"`—using a shallow copy so the caller’s options dictionary is unchanged. Rust Serde handling decodes only those exact sentinels.

This preserves backend-dependent validation:

- Native construction and successful-native `best_available` ignore all three regression-only weights.
- Regression construction and fallback `best_available` reject decoded nonfinite weights with the corresponding field-specific error.

Tuning now selects the maximum finite positive wall time in seconds before converting it to milliseconds. Native engine evaluation therefore still occurs when that conversion overflows, preserving Native error ordering, while Regression rejects the nonfinite target and retains no observation.

The public Rust/PyO3/Python APIs, type stubs, regression design document, architecture documentation, and focused tests are updated to reflect these contracts.

### AISimulate integration

The migrated implementation keeps the AIC mirror under `crates/core/src/perfmodel` and `python/aisimulate`, while preserving AISimulate’s Replay-owned `crate::engine` namespace and Python feature gates. The synchronization ledger advances from `ce2824e8` to source head `c8aee02f`; the manual report records why standalone AIC version-only manifest and lockfile changes are not copied into the unified AISimulate artifacts.

## Compatibility and deferred work

This is a documented next-minor source break:

- `from_regression` and fallback-capable `best_available` constructors now require `worker_type`.
- Adding the three public option fields breaks exhaustive Rust `ForwardPassPerfOptions` literals; downstream callers should use `..Default::default()`.

Strict-native constructor signatures, diagnostic wire shapes, FPM/engine schemas, and package versions are unchanged.

Dynamo worker-role forwarding remains pending downstream integration. Retaining original per-rank FPM observations for replay or a future Gaussian-process backend is deferred. The previously identified extreme finite feature/target fitting-overflow behavior is intentionally unchanged.

## Validation

- `cargo test --workspace`
- `cargo test -p aisimulate-core --lib perfmodel::fpm::` — 62 passed
- `cargo test -p aisimulate-core --features python --lib perfmodel::fpm::` — 62 passed
- `cargo test --manifest-path crates/tests/public-api/Cargo.toml` — 4 passed
- Built the combined `aisimulate` wheel with Maturin and tested the extracted artifact
- Raw/cross-package Python public-API suite — 28 passed
- Full SDK engine-step suite — 48 passed
- AIC synchronization tests — 3 passed
- CODEOWNERS policy tests — 99 passed
- Ruff check and format checks on changed Python files
- `cargo fmt --all -- --check`
- `git diff --check`

## Reviewer guide

The design contract is in `python/aisimulate/docs/fpm/aic-fpm-regression-design.md`. The main review paths are:

1. `crates/core/src/perfmodel/fpm/model.rs` for role-specific feature extraction and tuning ordering.
2. `crates/core/src/perfmodel/fpm/regression.rs` for retention, standardization, and fitting.
3. `crates/core/src/perfmodel/fpm/options.rs` and `python/aisimulate/src/aiconfigurator_core/sdk/rust_engine_step.py` for weight validation and JSON transport.
4. `docs/aic-sync-ce2824e-to-c8aee02-manual.md` for repository-transfer provenance and target-specific adaptations.

### CODE REVIEW

File: python/aisimulate/docs/fpm/aic-fpm-regression-design.md [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+## Dynamo integration and deferred work
+
+After a matching next-minor core release, Dynamo should pass its existing
+engine `worker_type` through both `from_regression` and `best_available`, expose
+the three weights as positive finite Planner configuration knobs with unity
+defaults, and include them in the engine-model reconstruction key. Weight
+changes require reconstruction followed by replay of Dynamo's retained raw FPM
+iterations. The role remains a constructor argument rather than an options-map
+entry, and Dynamo must update its exact core dependency pins together. This
+worktree changes only AIC: forwarding the worker role from Dynamo remains
```
Comment (jasonqinzhou) [inline python/aisimulate/docs/fpm/aic-fpm-regression-design.md]: This sentence still describes the source AIC worktree (`changes only AIC`), but this document now ships from AISimulate and this PR records the final AIC-to-AISimulate transfer. Could we reword it to something like `This change updates only AISimulate's imported AIC forward-pass model; forwarding the worker role from Dynamo remains pending downstream integration.`? That keeps the repository ownership boundary accurate for readers of the new implementation contract.

**Review by jasonqinzhou — APPROVED** [review]
**APPROVE · 8.8/10 · high confidence**

## What this PR does

This PR completes the frozen AIConfigurator-to-AISimulate transfer of the forward-pass regression redesign. It binds regression models to an immutable prefill, decode, or aggregated worker role; replaces inferred per-workload stores with a standardized two-feature fit and logarithmic retention grid; and carries the public Rust/PyO3/Python API, documentation, tests, and synchronization ledger forward together.

The change is unusually careful about preserving the native AIC path while isolating regression-only validation, and its tests cover role compatibility, cross-rank feature composition, numeric edge cases, JSON sentinel transport, and the external Rust API.

## Why this score

The 8.8 score reflects the completed Codex code-quality assessment and passing local verification, which found only one non-blocking documentation issue. The packet recommendation is nevertheless BLOCKED because the mandatory independent helper requested `claude-fable-5` but observed only `claude-haiku-4-5-20251001` and `claude-sonnet-4-6`; under the dual-review safety contract, that output cannot be represented as a completed Fable review or used to authorize comments.

File: python/aisimulate/docs/fpm/aic-fpm-regression-design.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+| \(B_d\) | `num_decode_requests` | scheduled Decode requests |
+| \(K_d\) | `sum_decode_kv_tokens` | Decode KV tokens read |
+
+The telemetry schema permits fully cached metadata with \(P_d=0\) and
+\(H_d>0\). Cached Prefill KV creates attention work only when fresh Prefill
+work exists:
+
+\[
+\widetilde H_d=H_d\,\mathbf 1[P_d>0].
+\]
```
Comment (tedzhouhk) [inline python/aisimulate/docs/fpm/aic-fpm-regression-design.md]: looks like the latex is not rendered, maybe try $$ ... $$?

File: python/aisimulate/docs/fpm/aic-fpm-regression-design.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+\left[
+\max_d\left(\alpha(\widetilde H_d+K_d)+\beta Q_d\right),
+\ \gamma\sum_d(P_d+B_d)
+\right].
+\]
+
+The Aggregated maximum is taken after composing the entire rank-local
+attention score; the implementation must not combine independent maxima from
+different ranks. Counters are converted to `f64` before multiplication or
+cross-rank summation, and derived features must be finite and nonnegative.
```
Comment (tedzhouhk) [inline python/aisimulate/docs/fpm/aic-fpm-regression-design.md:118]: Could we add a compact table here showing the exact feature vector for each worker type at attention_dp=1 and attention_dp>1? The equations above are correct, but putting the single-rank and multi-rank reductions side by side would make two important facts much easier to see: the regression always remains two-dimensional, and DP ranks are reduced with max for critical attention and sum for global FFN/MoE work.

| Worker | attention_dp=1 | attention_dp>1 |
|---|---|---|
| Prefill | x[0] = α H̃ + β Q<br>x[1] = γ P | x[0] = max_d(α H̃_d + β Q_d)<br>x[1] = γ Σ_d P_d |
| Decode | x[0] = α K<br>x[1] = γ B | x[0] = α max_d(K_d)<br>x[1] = γ Σ_d B_d |
| Aggregated | x[0] = α(H̃ + K) + β Q<br>x[1] = γ(P + B) | x[0] = max_d[α(H̃_d + K_d) + β Q_d]<br>x[1] = γ Σ_d(P_d + B_d) |

Could we also expand the semantic description of the three weights below the table?

- α scales the KV-token-related term shared by Prefill and Decode: H̃ for Prefill and K for Decode. It is not a Prefill-only weight.
- β scales only the Prefill attention-pair approximation Q. It does not represent Decode attention.
- γ scales global tokenwise FFN/MoE work: P for Prefill, B for Decode, and P+B for Aggregated.

It would also help to state explicitly that H, P, and K are per-rank sums. Q is already N times the balanced per-request attention-pair estimate,

Q = N[(H/N)(P/N) + (P/N)((P/N)+1)/2],

which simplifies to the formula above. Readers should not multiply Q by N again.

File: python/aisimulate/docs/fpm/aic-fpm-regression-design.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+Python: from_regression(worker_type, options=None)
+        best_available(config, worker_type, options=None)
+```
+
+`from_native` and `from_native_with_roots` remain role-free. Requiring
+`worker_type` in the regression-only and fallback-capable constructors is a
+documented next-minor API break. Adding the three public regression-weight
+fields is also a Rust source break for exhaustive `ForwardPassPerfOptions`
+struct literals; downstream callers should use
+`ForwardPassPerfOptions { bucket_count: 8, ..Default::default() }`. The raw
```
Comment (tedzhouhk) [inline python/aisimulate/docs/fpm/aic-fpm-regression-design.md]: [P2] The documented bucket_count example is guaranteed to fail

ForwardPassPerfOptions { bucket_count: 8, ..Default::default() } cannot construct a model because validate_options requires bucket_count to be a perfect square and returns "bucket_count must be a perfect square" for 8. The same invalid example also appears in docs/core-api.md. Please use 16 (the documented default) or another perfect square such as 9 in both places.

**Review by tedzhouhk — APPROVED** [review]
Implementation matches the documented worker-role and regression design. The Rust, PyO3, and Python paths are consistent, and the focused regression, public API, formatting, and sync validations pass. Approving with one non-blocking P2 documentation correction for the invalid bucket_count example.

File: python/aisimulate/docs/fpm/aic-fpm-regression-design.md (hunk 1 above)
Comment (diffractiongoo) [reply python/aisimulate/docs/fpm/aic-fpm-regression-design.md]: Addressed in `de0b4ed1`. The boundary now states that this change updates only AISimulate's imported AIC forward-pass model and that Dynamo worker-role forwarding remains pending downstream integration. I also corrected the document status from `aiconfigurator-core` to `aisimulate-core`.

File: python/aisimulate/docs/fpm/aic-fpm-regression-design.md (hunk 2 above)
Comment (diffractiongoo) [reply python/aisimulate/docs/fpm/aic-fpm-regression-design.md]: Addressed in `de0b4ed1`. I converted all inline and display math in the design document to GitHub-supported `$...$` and `$$...$$` delimiters and made the same correction in the PR summary. Both were checked with GitHub's Markdown renderer.

File: python/aisimulate/docs/fpm/aic-fpm-regression-design.md (hunk 3 above)
Comment (diffractiongoo) [reply python/aisimulate/docs/fpm/aic-fpm-regression-design.md:118]: Addressed in `de0b4ed1`. The document now includes the requested `attention_dp = 1` versus `attention_dp > 1` feature table and explicit semantics for `alpha`, `beta`, and `gamma`. It also states that `H_d` and `P_d` are Prefill-request sums on rank `d`, `K_d` is the rank's Decode-request sum, and `N_d`/`B_d` are per-rank request counts. The expanded derivation shows that `Q_d = N_d[(H_d/N_d)(P_d/N_d) + (P_d/N_d)(P_d/N_d+1)/2]` simplifies to the implemented formula, so `Q_d` already totals the `N_d` Prefill requests on rank `d` and must not be multiplied by `N_d` again. It also distinguishes `N_d` from the attention-DP size and states that DP ranks are reduced afterward by the role-specific maximum and sum.

File: python/aisimulate/docs/fpm/aic-fpm-regression-design.md (hunk 4 above)
Comment (diffractiongoo) [reply python/aisimulate/docs/fpm/aic-fpm-regression-design.md]: Addressed in `de0b4ed1`. Both `python/aisimulate/docs/fpm/aic-fpm-regression-design.md` and `docs/core-api.md` now use the documented default `bucket_count: 16`, which satisfies the perfect-square requirement.
